pkg/ring/ring.go

```diff
@@ -588,3 +658,45 @@ func (r *Ring) HasInstance(instanceID string) bool {
 	_, ok := instances[instanceID]
 	return ok
 }
+
+func (r *Ring) getCachedShuffledSubring(identifier string, size int) *Ring {
+	r.mtx.RLock()
+	defer r.mtx.RUnlock()
+
+	if r.lastTopologyChange.IsZero() || r.lastRingChange.IsZero() {
+		// Don't do any caching if this ring has no updates yet.
+		return nil
+	}
+
+	cached := r.shuffledSubringCache[identifier]
+	if cached == nil {
+		return nil
+	}
+
+	cached.mtx.Lock()
+	defer cached.mtx.Unlock()
+
+	if cached.ringShardSize != size {
```

Comment:
Good idea. I've opted to use struct instead of string as cache key instead of string, as that is [faster and uses no allocations based on benchmark](https://gist.github.com/pstibrany/7aaeeec2ccd6ce8b03111ba4585c849c).